streambed-machine/src/lib.rs

```diff
@@ -0,0 +1,275 @@
+use async_stream::stream;
+use futures_util::{Stream, StreamExt};
+use rand::thread_rng;
+use serde::{de::DeserializeOwned, Serialize};
+use std::{future::Future, marker::PhantomData, pin::Pin, vec::Vec};
+use streambed::{
+    commit_log::{Offset, ProducerRecord, Subscription, Topic},
+    decrypt_buf, encrypt_struct_with_secret, get_secret_value,
+    secret_store::SecretStore,
+};
+
+pub use streambed::commit_log::{CommitLog, ProducerError};
+
+/// Provides a key for an item that will be stored in the log record.
+pub trait RecordKey {
+    fn record_key(&self) -> u64;
+}
+
+/// Wraps a `CommitLog` and specializes it for a specific payload type.
+/// This adds the type, topic and the encoding and encryption scheme.
+#[derive(Debug)]
+pub struct LogAdapter<L, C, A> {
+    commit_log: L,
+    codec: C,
+    topic: Topic,
+    group: String,
+    marker: PhantomData<A>,
+}
+
+/// Provides a method on `CommitLog` to specialize it for a payload type.
+pub trait CommitLogExt
+where
+    Self: CommitLog + Sized,
+{
+    /// Specialize this commit log for items of type `A`
+    /// The topic and group names are given and a `Codec`
+    /// for encoding and decoding values of type `A`.
+    fn adapt<A>(
+        self,
+        topic: &str,
+        group: &str,
+        codec: impl Codec<A>,
+    ) -> LogAdapter<Self, impl Codec<A>, A> {
+        LogAdapter {
+            commit_log: self,
+            codec,
+            topic: topic.into(),
+            group: group.into(),
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<L> CommitLogExt for L where L: CommitLog {}
+
+impl<L, C, A> LogAdapter<L, C, A>
+where
+    L: CommitLog,
+    C: Codec<A>,
+    A: RecordKey + Clone + 'static,
+{
+    /// Send one item to the underlying commit log.
+    pub async fn produce(&self, item: A) -> Result<Offset, ProducerError> {
+        let key = item.record_key();
+        let topic = self.topic.clone();
+
+        if let Some(value) = self.codec.encode(item).await {
+            self.commit_log
+                .produce(ProducerRecord {
+                    topic,
+                    headers: Vec::new(),
```

Comment:
Would `vec!([])` be more efficient?

Reply:
Nah, check the docs: https://doc.rust-lang.org/std/vec/struct.Vec.html#method.new "The vector will not allocate until elements are pushed onto it."